### PR TITLE
[ADD] l10n_it E-invoicing.

### PR DESCRIPTION
**Pull request by jbm-odoo**
Generate E-invoice in xml, send E-invoice by PEC, manage mail (create new income invoice, manage response)

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/account/models/account_invoice_import_wizard.py [hunk 1]
```diff
@@ -11,11 +11,14 @@ def _get_default_journal_id(self):
         return self.env['account.journal'].search([('type', '=', 'purchase')], limit=1)
 
     attachment_ids = fields.Many2many('ir.attachment', string='Files')
-    journal_id = fields.Many2one(string="Journal", comodel_name="account.journal", required=True, domain="[('type', '=', 'purchase')]", default=_get_default_journal_id, help="Journal where to generate the bills")
```
Comment (jco-odoo) [inline addons/account/models/account_invoice_import_wizard.py]: @jbm-odoo Better to put the formatting in a separate commit.  Or only do it when you really change something.  Makes it difficult to check what changed.

File: addons/l10n_it/models/l10n_it_account_invoice.py [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+        ('delivered_accepted', 'This invoice is delivered and accepted by destinatory'),
+        ('delivered_refused', 'This invoice is delivered and refused by destinatory'),
+        ('delivered_expired', 'This invoice is delivered and expired (expiry of the maximum term for communication of acceptance/refusal)'),
+        ('failed_delivery', 'Delivery impossible, ES certify that it has received the invoice and that the file \
+                        could not be delivered to the addressee') # ok we must do nothing
+    ], default='to_send')
+
+    stamp_duty = fields.Float(default=0, string="Dati Bollo", size=15, readonly=True, states={'draft': [('readonly', False)]})
+
+    transport_document_number = fields.Char(string="Numero DDT", size=20, readonly=True, states={'draft': [('readonly', False)]}, help="Transport document number")
```
Comment (jco-odoo) [inline addons/l10n_it/models/l10n_it_account_invoice.py]: No way to just do a many2many between ddts and invoices and compute the dates from there?

Comment (jco-odoo) [reply addons/l10n_it/models/l10n_it_account_invoice.py]: Actually, a ddt can just be linked with a many2one to a picking, no need for many2many and everything could be calculated.

File: addons/l10n_it/models/l10n_it_account_invoice.py [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+    _inherit = "account.tax"
+
+    VAT_due_date = fields.Selection([
+        ("I", "[I] IVA ad esigibilità immediata"),
+        ("D", "[D] IVA ad esigibilità differita"),
+        ("S", "[S] Scissione dei pagamenti")], default="I", string="VAT due date")
+
+    has_exoneration = fields.Boolean(string="Has exoneration of tax", help="Tax has a tax exoneration.")
+    kind_exoneration = fields.Char(string="Exoneration", size=2, help="Exoneration type: First letter must be a N, second must be a integer. For example: N1 or N6", default="N1")
+    law_reference = fields.Char(string="Law Reference", size=100, default="Law ...")
```
Comment (jco-odoo) [inline addons/l10n_it/models/l10n_it_account_invoice.py]: Maybe better to use a placeholder in the view than a default value that makes no sense.

File: addons/l10n_it/models/l10n_it_account_invoice.py (hunk 2 above)
Comment (jco-odoo) [reply addons/l10n_it/models/l10n_it_account_invoice.py]: So, ddt here could just be a manual field, which we inherit when stock is installed.

File: addons/l10n_it/models/l10n_it_account_invoice.py [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+    transport_document_date = fields.Date(string="Data DDT", readonly=True, states={'draft': [('readonly', False)]}, help="Transport document date")
+
+    document_order_data_line_ids = fields.One2many('document.order.data', 'invoice_id', string='Document order',
+                                                   readonly=True, states={'draft': [('readonly', False)]}, copy=True)
+
+    ddt_id = fields.Many2one('l10n.it.ddt', string='DDT', copy=True)
+
+    ddt_ids = fields.Many2many('l10n.it.ddt', string='DDT', compute='_compute_ddt_ids')
+
+    document_unique_seq = fields.Char(help="Unique sequence use to send this document to government", readonly=True)
```
Comment (jco-odoo) [inline addons/l10n_it/models/l10n_it_account_invoice.py]: So the standard sequence not fine?

**Comment by nhomar** [comment]
won't you do this case l10n_it_edi?

IMHO mix data module with functional module will be more problematic than helpful. (when include multilocalization this will be so messy).

File: addons/l10n_it/models/account_invoice.py [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+    ], default='to_send')
+
+    stamp_duty = fields.Float(default=0, string="Dati Bollo", size=15, readonly=True, states={'draft': [('readonly', False)]})
+
+    document_order_data_line_ids = fields.One2many('document.order.data', 'invoice_id', string='Document order',
+                                                   readonly=True, states={'draft': [('readonly', False)]}, copy=True)
+
+    ddt_id = fields.Many2one('l10n.it.ddt', string='DDT', readonly=True, states={'draft': [('readonly', False)]})
+
+    document_unique_seq = fields.Char(help="Unique sequence use to send this document to government", readonly=True)
```
Comment (antonylesuisse) [inline addons/l10n_it/models/account_invoice.py]: l10n_it_progressive_number

File: addons/l10n_it/models/account_invoice.py [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+
+        if not self.company_id.mail_pec_server_id or not self.company_id.address_send_fatturapa:
+            for invoice in self:
+                invoice.message_post(
+                    body=(_("Error when sending mail with E-Invoice: Your company must have a mail PEC server and must indicate the mail PEC that will send electronic invoice."))
+                    )
+                invoice.send_state = ('invalid')
+            return
+
+        attachment = self.env['ir.attachment'].search([
```
Comment (antonylesuisse) [inline addons/l10n_it/models/account_invoice.py]: use binary field

File: addons/l10n_it/models/account_invoice.py [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+                invoice.message_post(
+                    body=(_("Error when sending mail with E-Invoice: %s") % (error.args[0]))
+                    )
+                invoice.send_state = ('invalid')
+
+class AccountInvoiceLine(models.Model):
+    _inherit = 'account.invoice.line'
+    _name = 'account.invoice.line'
+
+    ddt_line_ids = fields.Many2many('l10n.it.ddt', string='DDT', compute='_compute_ddt_ids')
```
Comment (antonylesuisse) [inline addons/l10n_it/models/account_invoice.py]: remove atm

File: addons/l10n_it/models/account_invoice.py [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+    def _compute_ddt_ids(self):
+        for line in self:
+            line.ddt_line_ids = line.invoice_id.ddt_id
+
+
+class AccountTax(models.Model):
+    _name = "account.tax"
+    _inherit = "account.tax"
+
+    VAT_due_date = fields.Selection([
```
Comment (antonylesuisse) [inline addons/l10n_it/models/account_invoice.py]: l10n_it_vat_due_date

File: addons/l10n_it/models/account_invoice.py [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+        ("I", "[I] IVA ad esigibilità immediata"),
+        ("D", "[D] IVA ad esigibilità differita"),
+        ("S", "[S] Scissione dei pagamenti")], default="I", string="VAT due date")
+
+    has_exoneration = fields.Boolean(string="Has exoneration of tax", help="Tax has a tax exoneration.")
+    kind_exoneration = fields.Char(string="Exoneration", size=2, help="Exoneration type: First letter must be a N, second must be a integer. For example: N1 or N6", default="N1")
```
Comment (antonylesuisse) [inline addons/l10n_it/models/account_invoice.py]: Selection '', N1 N2 N3

File: addons/l10n_it/models/account_invoice.py [hunk 10]
```diff
@@ -0,0 +1,407 @@
...
+                if record.kind_exoneration == 'N6' and record.VAT_due_date == 'S':
+                    raise UserError(_("'Scissione dei pagamenti' is not compatible with exoneration of kind 'N6'"))
+            if not record.has_exoneration and record.amount == 0:
+                raise ValidationError("Is the tax has an amount of 0.0, you must indicate the kind of exoneration.")
+
+class AccountPaymentTerm(models.Model):
+    _name = "account.payment.term"
+    _inherit = "account.payment.term"
+
+    payment_method = fields.Selection(selection=PAYMENT_METHODS,
```
Comment (antonylesuisse) [inline addons/l10n_it/models/account_invoice.py]: Not required ?

File: addons/l10n_it/models/ddt.py [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+
+
+class L10nItDdt(models.Model):
+    _name = 'l10n.it.ddt'
+
+    invoice_id = fields.One2many('account.invoice', 'ddt_id', string='Invoice Reference',
+                                 ondelete='cascade')
+
+    name = fields.Char(string="Numero DDT", size=20, help="Transport document number", required=True)
+    transport_document_date = fields.Date(string="Data DDT", help="Transport document date", required=True)
```
Comment (antonylesuisse) [inline addons/l10n_it/models/ddt.py]: date

File: addons/l10n_it/models/account_invoice.py [hunk 12]
```diff
@@ -0,0 +1,407 @@
...
+            if attachment and invoice.send_state not in ['invalid', 'to_send']:
+                raise UserError(_("You can't regenerate an E-Invoice when the first one is sent and there are no errors"))
+            if attachment:
+                attachment.unlink()
+            report_name = '%(country_code)s%(codice)s_%(progressive_number)s.xml' % {
+                'country_code': invoice.company_id.country_id.code,
+                'codice': invoice.company_id.codice_fiscale,
+                'progressive_number': invoice.document_unique_seq,
+                }
+            self.doc_unique_name = report_name
```
Comment (jco-odoo) [inline addons/l10n_it/models/account_invoice.py]: self -> invoice

File: addons/l10n_it/models/account_invoice.py [hunk 13]
```diff
@@ -0,0 +1,407 @@
...
+    def _check_exoneration_with_no_tax(self):
+        for record in self:
+            if record.has_exoneration:
+                if not record.kind_exoneration or not record.law_reference or record.amount != 0:
+                    raise ValidationError("If the tax has exoneration, you must enter a kind of exoneration, a law reference and the amount of the tax must be 0.0.")
+                if len(record.kind_exoneration) != 2 or record.kind_exoneration[0] != 'N' or not record.kind_exoneration[1].isdigit():
+                    raise ValidationError("The format of the kind of exoneration is not correct. Example of correct format: 'N1' or 'N6' ")
+                if record.kind_exoneration == 'N6' and record.VAT_due_date == 'S':
+                    raise UserError(_("'Scissione dei pagamenti' is not compatible with exoneration of kind 'N6'"))
+            if not record.has_exoneration and record.amount == 0:
```
Comment (jco-odoo) [inline addons/l10n_it/models/account_invoice.py]: either remove, either not record.child_tax_ids (not sure about the field)

File: addons/l10n_it/models/account_invoice.py [hunk 14]
```diff
@@ -312,96 +292,102 @@ def _create_zip(self, attachments):
...
+            'email_to': self.env.user.company_id.l10n_it_address_recipient_fatturapa,
         })
         try:
             mail_fattura.send(raise_exception=True)
             for invoice in self:
                 invoice.message_post(
                     body=(_("Mail sent on %s by %s") % (fields.Datetime.now(), self.env.user.display_name))
                     )
-                invoice.send_state = ('sent')
+                invoice.l10n_it_send_state = ('sent')
```
Comment (jco-odoo) [inline addons/l10n_it/models/account_invoice.py]: why (?  invoice.l10n_it_send_state = 'sent'
